tangelo/toolboxes/molecular_computation/tests/test_rdms.py

```diff
@@ -131,22 +133,50 @@ def test_compute_rdms_from_classical_shadow(self):
         assert_allclose(rdm1ssr, rdm1ss, atol=0.05)
         assert_allclose(rdm2ssr, rdm2ss, atol=0.05)
 
-    def test_pad_rdms_with_frozen_orbitals(self):
-        """Test padding of RDMs with frozen orbitals indices."""
+    def test_pad_restricted_rdms_with_frozen_orbitals(self):
+        """Test padding of RDMs with frozen orbitals indices (restricted)."""
 
         mol = mol_H2O_sto3g.freeze_mos([0, 3, 4, 5], inplace=False)
 
         onerdm_to_pad = np.loadtxt(pwd_this_test + "/data/H2O_sto3g_onerdm_frozen0345.data")
         twordm_to_pad = np.loadtxt(pwd_this_test + "/data/H2O_sto3g_twordm_frozen0345.data")
 
-        test_onerdm, test_twordm = pad_rdms_with_frozen_orbitals(mol, onerdm_to_pad, twordm_to_pad.reshape((3,)*4))
+        test_onerdm, test_twordm = pad_rdms_with_frozen_orbitals_restricted(mol, onerdm_to_pad, twordm_to_pad.reshape((3,)*4))
 
         padded_onerdm = np.loadtxt(pwd_this_test + "/data/H2O_sto3g_padded_onerdm_frozen0345.data")
         padded_twordm = np.loadtxt(pwd_this_test + "/data/H2O_sto3g_padded_twordm_frozen0345.data")
 
         np.testing.assert_array_almost_equal(padded_onerdm, test_onerdm, decimal=3)
         np.testing.assert_array_almost_equal(padded_twordm.reshape((7,)*4), test_twordm, decimal=3)
 
+    def test_pad_unrestricted_rdms_with_frozen_orbitals(self):
+        """Test padding of RDMs with frozen orbitals indices (unrestricted)."""
+
+        mol = SecondQuantizedMolecule(mol_H2O_sto3g.xyz, uhf=True, frozen_orbitals=[(0, 3, 4, 5), (0, 3, 4, 5)])
+
+        onerdm_a_to_pad = np.loadtxt(pwd_this_test + "/data/H2O_UHF_sto3g_onerdm_frozen0345_alpha.data")
```

Comment:
I am usually not fond of having entire matrices loaded and compared terms by terms in tests.
That usually makes the tests pretty sensitive to changes in the compute package that uses them, stores a large quantity of data (mostly zeros in your case).

If you think this is appropriate and the precision of your assert statement is good enough, then it's alright.
The alternative is to instead make assert statements on properties of these data-structures (their trace? norm? determinant ? sparse structure ? the value of an expression that uses them ?) that ideally has some physical / practical meaning. Right now, your test will simply tell you whether entry by entry, you are close enough to what PYSCF or whatever reference tool you used computed at a given time. If not, well all you know is that the entries aren't the same (but maybe its good, and these guys fixed a bug or something).

For example, the 1 and 2-rdm tests with DMET revealed that computing an expression using the trace of the matrix should match the number of electrons of the system. What's great about this is that regardless of the package we use and the sensitivity of the calculations / bugs in there, is that it focuses on values that truly matter instead of arbitrary quantities that may be meaningless to the reader.

It is up to you and your creativity here.

Reply:
The whole purpose of this function is to pads the RDMs with (2s/1s (depending on UHF or RHF) or 0s). I designed the test so the padding is not trivial (not only the very fist or very last orbitals), so that the padding happens "inside" the numpy arrays.

Evaluating global quantities (trace, norm, determinant) in this situation does not make sense in my opinion, because a misplaced element in the RDMs can have a big impact on the computed eigenvalue (especially with DMET where it is used). Trace could be used to check if the number of electrons increases, but it would not tell if the off-diagonal are OKs.

Reply:
Alright then, we'll keep it that way :)

Another way to test this would be to just track the positions / number of zero and non-zero elements, and maybe check that the sum of all elements in the matrix does not change before and after padding (I'm not saying this is what we should do, just giving you an example of how creative we can be). Happy to proceed with what you have :)